pkg/tfbridge/info/external_methods.go

```diff
@@ -0,0 +1,109 @@
+// Copyright 2016-2024, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package info
+
+// This file maintains backward comparability by maintaining methods on *Provider that are
+// supplied by packages that import info.
+//
+// It should not be extended and new functions similar to those here should be
+// free-standing methods.
+
+import (
+	"strings"
+
+	// Needed to unsafely import ApplyAutoAliases
+	_ "unsafe"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
+)
+
+// Materialize [tfbridge.ApplyAutoAliases] as a method in [*Provider].
+//
+// This needs linkname to avoid go's package coherence rules.
+
+//go:linkname applyAutoAliases github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfbridge.ApplyAutoAliases
```

Comment:
Can you please elaborate a bit here on the usage of linkname?

Reply:
This links the local symbol `applyAutoAliases` to the symbol defined at `github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfbridge.ApplyAutoAliases`. We use this to allow `info` to link in a function from `/pkg/tfbridge`, even though it is illegal to import `/pkg/tfbridge`.

If your asking about how `go:linkname` actually works, https://pkg.go.dev/cmd/compile is a better source of information then I am here.

Reply:
Yeah, that makes sense